modules/extensions/src/backend/index.ts

```diff
@@ -28,10 +33,97 @@ const onServerReady = async (bp: typeof sdk) => {
       res.sendStatus(404)
     })
   )
+
+  await setupMiddleware(bp)
+}
+
+const onModuleUnmount = async (bp: typeof sdk) => {
+  await removeMiddleware(bp)
+}
+
+async function setupMiddleware(bp: typeof sdk) {
+  bp.events.registerMiddleware({
+    description: 'Extends outgoing messages',
+    direction: 'outgoing',
+    handler: outgoingHandler,
+    name: MIDDLEWARE_NAME,
+    order: 10
+  })
+
+  async function outgoingHandler(event: sdk.IO.OutgoingEvent, next: sdk.IO.MiddlewareNextCallback) {
+    const logger = bp.logger.forBot(event.botId)
+    // logger.info(`outgoingHandler: ${JSON.stringify(event)}`)
+
+    if (event.payload && event.payload.options) {
+      events[event.target] = event
+    }
+
+    return next()
+  }
+
+  bp.events.registerMiddleware({
+    description: 'Extends incoming messages',
+    direction: 'incoming',
+    handler: incomingHandler,
+    name: MIDDLEWARE_NAME,
+    order: 0
+  })
+
+  function normalize(str: string) {
+    return str
+      .toLowerCase()
+      .normalize('NFD')
+      .replace(/[\u0300-\u036f]/g, '')
+      .split(/ \(|,/)[0]
+  }
+
+  async function incomingHandler(event: sdk.IO.IncomingEvent, next: sdk.IO.MiddlewareNextCallback) {
+    const logger = bp.logger.forBot(event.botId)
+    const outgoingEvent = events[event.target]
+
+    // logger.info(`incomingHandler: \nevent=${JSON.stringify(event)} \noutgoingEvent=${JSON.stringify(outgoingEvent)}`)
+
+    if (outgoingEvent) {
+      delete events[event.target]
+      const options = outgoingEvent.payload.options
+      const text = event.payload.text
+      // logger.info('options:', options, event.payload)
+
+      const option = _.filter(
+        options,
+        option => normalize(option.label) == normalize(text) || option.value.toLowerCase() == text.toLowerCase()
+      )
+
+      logger.info(`option: ${JSON.stringify(option)}`)
+
+      event.state.temp = _.assign(event.state.temp || {}, { 'dropdown-valid': true })
+
+      event.setFlag(bp.IO.WellKnownFlags.FORCE_PERSIST_STATE, true)
+
+      if (option.length) {
+        event.payload.text = option[0].label
+        event.payload.payload = option[0].value
+      } else {
+        delete event.payload.text
+
+        event.payload.invalid = text
+        event.state.temp['dropdown-valid'] = false
+
+        return next(new Error(`Invalid value: ${text}`))
+      }
+    }
+
+    return next()
+  }
+}
+
+async function removeMiddleware(bp: typeof sdk) {
+  bp.events.removeMiddleware(MIDDLEWARE_NAME)
```

Comment:
What are these changes meant for? If it's to fix the dropdown changing the if statements in `renderElement` function of modules/extensions/src/content-types/dropdown.js would have been enough. This seems overly complicated.

Reply:
We couldn't find a better way to ensure that the incoming messages after rendering the dropdown options indeed validate against the options provided by the dropdown in the SMS/WhatsApp flows. If you have a better idea how to fetch the dropdown options when the following incoming event is received to validate it, let us know and we can change it.